milestones/api.py

```diff
@@ -0,0 +1,325 @@
+    """ Validation helper """
+    if not validators.user_is_valid(user):
+        exceptions.raise_exception(
+            "User",
+            user,
+            exceptions.InvalidUserException
+        )
+
+
+# PUBLIC FUNCTIONS
+def get_milestone_relationship_types():
+    """
+    Exposes the available relationship type choices without exposing the data layer
+    """
+    return data.RELATIONSHIP_TYPES
+
+
+def add_milestone(milestone):
+    """
+    Passes a new milestone to the data layer for storage
+    """
+    _validate_milestone(milestone)
+    milestone = data.create_milestone(milestone)
+    return milestone
+
+
+def edit_milestone(milestone):
+    """
+    Passes an updated milestone to the data layer for storage
+    """
+    _validate_milestone(milestone)
+    return data.update_milestone(milestone)
+
+
+def get_milestone(milestone_id):
+    """
+    Retrieves the specified milestone
+    """
+    return data.fetch_milestone(milestone_id)
+
+
+def get_milestones(namespace):
+    """
+    Retrieves the specified milestone by namespace
+    (no other way to really do it right now)
+    """
+    milestone = {
+        'namespace': namespace,
+    }
+    _validate_milestone(milestone)
+    return data.fetch_milestones(milestone)
+
+
+def remove_milestone(milestone_id):
+    """
+    Removes the specified milestone
+    """
+    milestone = {
+        'id': milestone_id,
+    }
+    data.delete_milestone(milestone)
+
+
+def add_course_milestone(course_key, relationship, milestone):
+    """
+    Adds a course-milestone link to the system
+    'relationship': string value (eg: 'requires')
+    """
+    _validate_course_key(course_key)
+    _validate_milestone_relationship_type(relationship)
+    _validate_milestone(milestone)
+    data.create_course_milestone(course_key=course_key, relationship=relationship, milestone=milestone)
+
+
+def get_course_milestones(course_key, relationship=None):
+    """
+    Retrieves the set of milestones for a given course
+    'relationship': optional filter on milestone relationship type (string, eg: 'requires')
+    Returns an array of dicts containing milestones
+    """
+    _validate_course_key(course_key)
+
+    if relationship is not None:
+        _validate_milestone_relationship_type(relationship)
+    return data.fetch_courses_milestones(course_keys=[course_key], relationship=relationship)
+
+
+def get_course_required_milestones(course_key, user):
+    """
+    Retrieves the set of required milestones for a given course that a user has not yet collected
+    """
+    _validate_course_key(course_key)
+    _validate_user(user)
+    required_milestones = data.fetch_courses_milestones(
+        [course_key],
+        get_milestone_relationship_types()['REQUIRES'],
+        user
+    )
+    return required_milestones
+
+
+def get_course_milestones_fulfillment_paths(course_key, user):
+    """
+    Returns a collection composed of the possible fulfillment/collection opportunites
+    """
```

Comment:
Updated the docstring to include inputs and outputs -- pared-down the actual output container to only what is needed after seeing what it actually looked like.